genai-perf/genai_perf/goodput_calculator/llm_goodput_calculator.py

```diff
@@ -0,0 +1,183 @@
+            logger.info(
+                f"Invalid SLOs found: {', '.join(invalid_slos)}. "
+                f"The goodput will be -1. Valid SLOs are: {valid_slos_list}."
+            )
+            self._goodput = self.INVALID_GOODPUT
+
+    def _combine_requests_metric_values(self) -> None:
+        """
+        Combine values from the metrics that match with the valid SLOs at a
+        per request level.
+        """
+        if self.goodput == self.INVALID_GOODPUT:
+            return
+
+        if self._has_time_target:
+            time_names = [
+                self.get_slo_base_name(key) for key in self._valid_time_related_slos
+            ]
+            requests_time_metric_values = [
+                self._metric.data[name] for name in time_names
+            ]
+
+            self._combined_requests_time_metric_values = list(
+                zip(*requests_time_metric_values)
+            )
+
+        if self._has_throughput_target:
+            throughput_names = [
+                self.get_slo_base_name(key)
+                for key in self._valid_throughput_related_slos
+            ]
+            requests_throughput_metric_values = [
+                self._metric.data[name] for name in throughput_names
+            ]
+
+            self._combined_requests_throughput_metric_values = list(
+                zip(*requests_throughput_metric_values)
+            )
+
+    def _count_good_reqs(self) -> Optional[int]:
+        """Count the number of good requests according to SLOs."""
+        if self.goodput == self.INVALID_GOODPUT:
+            return None
+        target_time_metric_values = []
+        target_throughput_metric_values = []
+        if self._has_time_target:
+            num_of_requests = len(self._combined_requests_time_metric_values)
+            target_time_metric_values = list(self._valid_time_related_slos.values())
+        if self._has_throughput_target:
+            num_of_requests = len(self._combined_requests_throughput_metric_values)
+            target_throughput_metric_values = list(
+                self._valid_throughput_related_slos.values()
+            )
+
+        good_req_count = 0
+        for idx in range(num_of_requests):
+            is_good_request = True
+            request_time_metric_values: List[float] = []
+            request_throughput_metric_values: List[float] = []
+            if self._has_time_target:
+                request_time_metric_values = list(
+                    self._combined_requests_time_metric_values[idx]
+                )
+            if self._has_throughput_target:
+                request_throughput_metric_values = list(
+                    self._combined_requests_throughput_metric_values[idx]
+                )
+            for val, slo in zip(request_time_metric_values, target_time_metric_values):
+                if val > slo:
+                    is_good_request = False
+                    break
+            else:
```

Comment:
I think the indentation is off, so I'm surprised the code works. Where does this else clause belong to?

Reply:
It is a relatively rare usage. The `else` clause belongs to previous `for` loop in order to avoid unnecessary loops.
https://book.pythontips.com/en/latest/for_-_else.html#:~:text=for%20loops%20also%20have%20an%20else%20clause%20which%20most%20of%20us%20are%20unfamiliar%20with.%20The%20else%20clause%20executes%20after%20the%20loop%20completes%20normally.

Reply:
Interesting. Is there another way to do this? Something that's rare and unknown is not good for code readability. If there is *no* other way, then we could add formatting to separate this for/else into its own block and add a comment explaining that's happening. However, in this case, we're talking about the code execution path being unclear or misunderstood, so that's dangerous.

One way to do this could be to check if is_good_request is False.

Reply:
Updated